packages/typescript/base.js

```diff
@@ -0,0 +1,439 @@
+module.exports = {
+  rules: {
+    // Require that member overloads be consecutive
+    // https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/adjacent-overload-signatures.md
+    "@typescript-eslint/adjacent-overload-signatures": "error",
+
+    // Requires using either T[] or Array<T> for arrays
+    // https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/array-type.md
+    "@typescript-eslint/array-type": "error",
+
+    // Disallows awaiting a value that is not a Thenable
+    // https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/await-thenable.md
+    "@typescript-eslint/await-thenable": "error",
+
+    // Bans @ts-<directive> comments from being used unless @ts-expect-error, which requires description
+    // https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/ban-ts-comment.md
+    "@typescript-eslint/ban-ts-comment": [
+      "error",
+      {
+        "ts-expect-error": "allow-with-description",
+        "ts-ignore": "allow-with-description",
+        "ts-nocheck": true,
+        "ts-check": true,
+        minimumDescriptionLength: 10,
+      },
+    ],
+
+    // We're not concerned about ts-lint: comments as they're not used anywhere in our codebase
+    // https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/ban-tslint-comment.md
+    "@typescript-eslint/ban-tslint-comment": "off",
+
+    // Bans specific types from being used, e.g.: {}
+    // https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/ban-types.md
+    "@typescript-eslint/ban-types": "error",
+
+    // Ensures that literals on classes are exposed as fields to avoid additional overhead of creating a closure
+    // https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/class-literal-property-style.md
+    "@typescript-eslint/class-literal-property-style": ["error", "fields"],
+
+    // Enforces the usage of Record<K, V> type instead of { [key: K]: V }
+    // https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/consistent-indexed-object-style.md
+    "@typescript-eslint/consistent-indexed-object-style": ["error", "record"],
```

Comment:
There is a semantic difference between `{ [key: K]: V }` and `Record<K, V>` so I think both should be used.

`Record` implies (and has type check support) that this object contains a value for all possible keys.

This is why I think `Record` only makes semantic sense with literal types, and `Record<number, whatever>` is better replaced with `{ [key: number]: whatever }`

One more benefit of not using `Record` when the key is "generic" is that you can describe what that key represents. For example: `{ [userId: number]: string }`

Reply:
To address some points:

> There is semantic difference between `{ [key: K]: V }` and `Record<K, V>`

- Yeah, in that `type Record<Key, Value> = { [key: Key]: Value }`. I don't know what else you mean though.

> `Record` implies (and has type check support) that this object contains a value for all possible keys.

- I'm not sure what you mean. They behave the exact same way. `Record` is an alias to `{ [k: K]: V }` ([see this TypeScript Playground example](https://www.typescriptlang.org/play?#code/C4TwDgpgBAogHgQwLZgDYQIxQLxQN5QDaA1hCFAJYB2UARABYSqoD2tUAPnQO4sBOqACa0AugC4oAZ2B9qAcygBfANwAoVQGMWVaVEwT4yNJhz5VUKI2YsJtXgOEAaVYvWhIsRCnQAmUwCUILT5BAB4GJlZ2Ljt+IVpHKRl5AD41TW1dCB8DL2M-XDxzS0ibHjinF1UgA))

> One more benefit of not using `Record` when the key is "generic" is that you can describe what that key represents. For example: `{ [userId: number]: string }`

- Yes this is true, but if you name the `type` and your variables correctly, it's completely redundant, e.g.:

```ts
type UsersById = Record<number, User>;

const usersById = {
  "kau7s52gh9d": { /* ... */ },
};
```

An additional point to consider is when you're trying to exhaust union types in object keys:

```ts
type ActionName = "play" | "pause" | "restart" | "skipForward" | "skipBackward" | "next" | "previous";

//                                      would you know ▼ this syntax without writing `action: ActionName` and searching Google?
type ActionHandlersByName = { [action in ActionName]: Reducer };

// vs
type Actioname = Record<ActionName, Reducer>;
```

Reply:
I knew that `action: ActionName` was invalid, but I had no idea you could use `in`.

> I'm not sure what you mean. They behave the exact same way. Record is an alias to `{ [k: K]: V }`

It's a similar idea to what we talked about regarding `null` vs `undefined` as type options. 
They both represent "missing", but there is a semantic difference.

In a similar fashion a rule from the typescript style guide suggest using both syntaxes depending on the semantics of the object.

[Rule + explanation](https://www.notion.so/superbet/TypeScript-Guide-4269c8f8208748dca58032f382f6a881#35e4e58554714cb18b561367d02fc399)